### ISSUE TITLE
Features: Installs to User Home Directory

### ISSUE THREAD

**Issue by virtualstaticvoid** [devcontainers/spec#91]
Hello,

I'm writing a feature which requires writing configuration files to the home directory of the user within the container.

Looking at the source code of existing implemented features as a reference, such as the [`anaconda`][anaconda] feature, I encountered code such as the following, which has specific logic to determine the user (and thus the home directory location).

https://github.com/devcontainers/features/blob/a5fbdab3c096152dd3a318c0c23da9465815aa69/src/anaconda/install.sh#L29-L44

I wasn't able to find specific documentation in the [proposal][proposal] on a "contract" for feature developers, or details on a `USERNAME` environment variable being provided when executing the `install.sh` script. Furthermore, use of "magical" users such as `automatic`, `auto`, `vscode`, `node`, `codespace` or other (via `/etc/passwd`) to aid in the determination behaviour is something I wouldn't want to have to repeat in authoring my features.

Furthermore, the reliance on having one of these user exist in a given base box is tenuous and makes this logic quite fragile (and smelly).

Therefore I propose a better and surprisingly much simpler solution, inspired in part by how the `vagrant` shell provisioner works, as follows:

1. The process which is executing the `install.sh` script, has access to the given `devcontainer.json` file, which defines a [`remoteUser`][remoteuser] and thus can be used to provide the `USERNAME` environment variable to the script; sort of like a docker build argument, which would remove the need to have to infer the correct user.

2. Since the feature `install.sh` script is typically executed as `root`, unless the base image switches the user, files installed to the home directory need to be `chown`ed to make the files accessible to the user when the dev container is running. This can be solved by having an optional `devcontainer-feature.json` property, such as `installAs`, which is an enum type with values of `root` or `remoteUser`. When `remoteUser` is provided, `install.sh` is executed in the context of the user provided by the `remoteUser` property. This is similar in concept to how the [`privileged`][vagrant-privileged] setting of the `vagrant` shell provisioner alters the context. This does place a limitation on what the `install.sh` script can do, so authors should include an assertion for root if required for proper execution.

   https://github.com/devcontainers/features/blob/a5fbdab3c096152dd3a318c0c23da9465815aa69/src/terraform/install.sh#L35-L38

3. To aid troubleshooting, since the onus is on the `devcontainer.json` author to ensure alignment of the user(s) contained in the base image and the `remoteUser` provided, a check should be performed before installing features, to assert that the given `remoteUser` exists and a nice helpful message if it doesn't.

[anaconda]: https://github.com/devcontainers/features/blob/main/src/anaconda
[proposal]: https://github.com/devcontainers/spec/blob/main/proposals/devcontainer-features.md#authoring
[remoteuser]: https://github.com/devcontainers/spec/blob/main/docs/specs/devcontainerjson-reference.md
[vagrant-privileged]: https://www.vagrantup.com/docs/provisioning/shell

**Comment by Chuxel** [comment]
Yeah agree with this - In fact, https://github.com/devcontainers/spec/issues/75 proposes to bring in first class support for creating them as well. Part of that would be both the remoteUser and containerUser would be available as env vars in Features as well.  That way you can `su <user> -c` to operate fully in that context where needed.  Would that hit what you are getting at here?  

A first step could just be to add in the env vars though, I agree - FYI @chrmarti since we've been talking about this one. I'd love to drop all of the "automatic" logic since it generally works, but is not foolproof since the UID of the user could be something other than `1000`.

As a FYI, there's also groups created in some features to help manage privs for some of this, but that's generally due to not having #25.

**Comment by virtualstaticvoid** [comment]
> A first step could just be to add in the env vars though

Yeah, that would work. Feature authors can then assume their `install.sh` will always be executed as `root`, and if need be `su <user> ...` can be used to execute as the given user, and `remoteUser` and `containerUser` should both be provided as environment variables to the `install.sh` script.

**Comment by chrmarti** [comment]
You could update the existing home folders and the `/etc/skel` folder that is used when users are added:

E.g., adding a `~/.bar` config:
```sh
for home in /etc/skel /home/*
do
	su $(stat -c '%U' "$home") -c "echo foo >'$home/.bar'"
done
```

**Comment by chrmarti** [comment]
@Chuxel Makes sense, let's track passing the effective username to the feature scripts here.

**Comment by stuartleeks** [comment]
I imagine that a common use of these in feature installation scripts will be to write to the user home folder in the container. 
Could we include a `USER_HOME` variable that is set to `/root` for `root` user and `/home/$REMOTE_USER` otherwise? (Just thinking that it would save multiple feature implementations from including that logic)

**Comment by chrmarti** [comment]
Makes sense. Some suggest using `getent passwd $USER | cut -d: -f6` or `eval echo "~$USER"`, but these are not obvious.

We should probably include the home folder for both users. Also prefixing the variables with `_` will make sure they do not collide with feature options. Updating the proposal above.

**Comment by chrmarti** [comment]
Continuing in https://github.com/devcontainers/spec/pull/110.
